### ISSUE TITLE
`bundle install` doesn't install runtime dependencies from gemspec on JRuby

### ISSUE THREAD

**Issue by alexdowad** [bundler/bundler#4157]
...And here is why. The story begins innocently in `dsl.rb`, line 63:

```ruby
        gem spec.name, :path => path, :glob => glob, :platforms => Bundler::Dependency::REVERSE_PLATFORM_MAP[spec.platform]
```

Fair enough. We want to install all the runtime dependencies for the gem defined in the gemspec. Rather than adding them all explicitly, we just add a dependency on the gem itself. BUT! Look at the last part of that line...

```ruby
:platforms => Bundler::Dependency::REVERSE_PLATFORM_MAP[spec.platform]
```

What will that return if the gem author didn't specify a platform in the gemspec? To answer that, we need to take a detour into Rubygems itself. `specification.rb`, lines 125-142:

```ruby
@@default_value = {
    :authors                   => [],
    :autorequire               => nil,
    :bindir                    => 'bin',
    :cert_chain                => [],
    :date                      => TODAY,
    :dependencies              => [],
    :description               => nil,
    :email                     => nil,
    :executables               => [],
    :extensions                => [],
    :extra_rdoc_files          => [],
    :files                     => [],
    :homepage                  => nil,
    :licenses                  => [],
    :metadata                  => {},
    :name                      => nil,
    :platform                  => Gem::Platform::RUBY,
```

So specifying no platform at all is the same as `spec.platform = 'ruby'`. And `Bundler::Dependency::REVERSE_PLATFORM_MAP['ruby']` is...

```ruby
[:ruby, :ruby_18, :ruby_19, :ruby_20, :ruby_21, :ruby_22, :ruby_23, :mri, :mri_18, :mri_19, :mri_20, :mri_21, :mri_22, :mri_23, :rbx]
```

So as far as Bundler is concerned, putting no platform at all in the gemspec is the same as saying "this is only for MRI and RBX"!

I can see three ways to solve this:

1. Amend `CurrentRuby` as follows:

```ruby
  def ruby?
    true
  end
```

The logic behind this is that saying a gem's platform is "ruby" simply means "this is a pure-Ruby gem, it doesn't have any platform dependencies". So that means the gem should be good on ALL Rubies. Right? Wrong? Maybe?

2. Special-case 'ruby' in `REVERSE_PLATFORM_MAP`.

3. Special-case platform 'ruby' in `Dsl#gemspec`, so it doesn't add the `:platforms => ...` option at all when `spec.platform` is `'ruby'`.

Comments please!

**Comment by Kris-LIBIS** [comment]
I'm not familiar with Bundler code, so I'll skip commenting on the fix you propose. I do want to upvote this issue as it breaks any travis jruby build for pure ruby gems that do not specify a platform by definition. Could you at least please revoke the released gem until this is fixed? All my travis builds are failing right now because of this, unless I start adding the runtime gems to the Gemfile for each gem.

**Comment by segiddins** [comment]
We don't yank gems. In the meantime, you're more than welcome to use 1.10.6. We had over two weeks of prereleases where nobody reported any issues, so be patient as we work to fix this in our spare time.

**Comment by alexdowad** [comment]
@segiddins, I'm happy to write and test the fix, but I'd like to hear what you think of the proposed 3 possible approaches above.

**Comment by segiddins** [comment]
We've actually backed out this change on stable already and @indirect is working on a fix as we speak!

**Comment by jkutner** [comment]
@segiddins @indirect thanks for getting on this so quickly! it's awesome that jruby users can rely on bundler to work. this is the first problem I've had in a very long time <3 <3

**Comment by Kris-LIBIS** [comment]
@segiddins don't get me wrong. I do appreciate the hard work you put into this marvelous gem and I have been relying on it for quite a few years now without a problem. As for using the older version, I'm not sure if we have that option on Travis and I think they skipped the pre-release versions because the problem appeared only a couple of days ago.

To be honest I did panic a bit as I did not have a workaround immediately and failing a build because of bundler obfuscates any issues with the build itself defying the purpose of it all. Since I added the platform line to my gemspecs, I'm as happy as I've ever been ;)

Keep up the good work and thanks for looking at this.
